File: src/Services/ArchT.Services.CustomTrade.Core/Contracts/EventContracts.fs
module EventContracts

open System

type EventContract() as this =
    let time = DateTime.UtcNow
    member val EventName=this.GetType().Name with get, set
    member val EventTime=time with get, set
    member val EventSequence=time.Ticks with get, set

type OrderLineUpdated(productId, quantity, price) =
    inherit EventContract()
    member val ProductId=productId with get, set
    member val Quantity=quantity with get, set
    member val Price=price with get, set
    new() = OrderLineUpdated("", 0, 0M)

type OrderUpdated(id:string, items:OrderLineUpdated list, processedTime:Option<DateTime>) =
    inherit EventContract()  
    member val Id=id with get, set
    member val Lines=items with get, set
    member val Processed=processedTime with get, set
    new() = OrderUpdated("", [], None)

File: src/Services/ArchT.Services.Search/Contracts/EventContracts.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ArchT.Services.Search.Contracts
{
    public abstract class EventContract
    {
        private DateTime? time;
        private long? sequence;
        private string name;
        public string EventName { get => string.IsNullOrEmpty(name) ? GetType().Name : name; set => name = value; }
        public DateTime EventTime { get => time ?? DateTime.UtcNow; set => time = value; }
        public long EventSequence { get => sequence ?? EventTime.Ticks; set => sequence = value; }
    }

    public class ProductUpdated : EventContract
    {
        public string ProductId { get; set; }
        public string Name { get; set; }
        public int Stock { get; set; }
    }
}
